datamol/utils/fs.py

```diff
@@ -1,20 +1,347 @@
+"""The `fs` module makes it easier to work with all type of path (the ones supported by `fsspec`).
+"""
+
 from typing import Union
+from typing import Optional
+from typing import List
 
 import os
+import io
+import hashlib
+import pathlib
 
 import fsspec
 
 
+def _import_tqdm():
+    try:
+        from tqdm.auto import tqdm
+
+        return tqdm
+    except ImportError:
+        return None
+
+
+def _import_appdirs():
+    try:
+        import appdirs
+
+        return appdirs
+    except ImportError:
+        return None
+
+
+def get_cache_dir(app_name: str, suffix: str = None, create: bool = True):
+    """Get a local cache directory for a given application name.
+
+    Args:
+        app_name: The name of the application.
+        suffix: A subdirectory appended to the cache dir.
+        create: Whether to create the directory and its parents if it does not
+            already exist.
+    """
+
+    appdirs = _import_appdirs()
+
+    if appdirs is None:
+        raise ImportError(
+            "To use `dm.utils.fs.get_cache_dir()`, you must have `appdirs` "
+            "installed: `conda install appdirs`."
+        )
+    cache_dir = pathlib.Path(appdirs.user_cache_dir(appname=app_name))
+
+    if suffix is not None:
+        cache_dir /= suffix
+
+    if create:
+        cache_dir.mkdir(exist_ok=True, parents=True)
+
+    return cache_dir
+
+
+def get_mapper(path: Union[str, os.PathLike]):
+    """Get the fsspec mapper.
+
+    Args:
+        path: a path supported by `fsspec` such as local, s3, gcs, etc.
+    """
+    return fsspec.get_mapper(path)
+
+
+def get_basename(path: Union[str, os.PathLike]):
+    """Get the basename of a file or a folder.
+
+    Args:
+        path: a path supported by `fsspec` such as local, s3, gcs, etc.
+    """
+    path = str(path)
+    mapper = get_mapper(path)
+    clean_path = path.rstrip(mapper.fs.sep)
+    return str(clean_path).split(mapper.fs.sep)[-1]
+
+
+def get_extension(path: Union[str, os.PathLike]):
+    """Get the extension of a file.
+
+    Args:
+        path: a path supported by `fsspec` such as local, s3, gcs, etc.
+    """
+    basename = get_basename(path)
+    return basename.split(".")[-1]
+
+
+def exists(path: Union[str, os.PathLike, fsspec.core.OpenFile, io.IOBase]):
+    """Check whether a file or a directory exists.
+
+    Important: HTTP file and file-like object always exists.
+
+    Args:
+        path: a path supported by `fsspec` such as local, s3, gcs, etc.
+    """
+    return is_file(path) or is_dir(path)
+
+
+def is_file(path: Union[str, os.PathLike, fsspec.core.OpenFile, io.IOBase]):
+    """Check whether a file exists.
+
+    Important: HTTP file and file-like object always exists.
+
+    Args:
+        path: a path supported by `fsspec` such as local, s3, gcs, etc.
+    """
+    if isinstance(path, fsspec.core.OpenFile):
+        if isinstance(path.fs.protocol, str) and "http" in path.fs.protocol:
+            # Assume HTTP to always exists.
```

Comment:
content type or head request ?

Reply:
It seems like the specific http logic is not required anymore. I also bumped the `fsspec` lower bound.